pkg/planner/core/indexmerge_path.go

```diff
@@ -478,6 +632,185 @@ func (ds *DataSource) generateIndexMergeAndPaths(normalPathCnt int) *util.Access
 	return indexMergePath
 }
 
+/*
+select * from t where ((1 member of (a) and b=1) or (2 member of (a) and b=2)) and (c > 10)
+
+	IndexMerge(OR)
+		IndexRangeScan(a, b, [1 1, 1 1])
+		IndexRangeScan(a, b, [2 2, 2 2])
+		Selection(c > 10)
+			TableRowIdScan(t)
+
+Two limitations now:
+1). Not support the embedded index merge case, which (DNF_Item1 OR DNF_Item2), for every DNF item,
+try to map it into a simple normal index path or mv index path, other than an internal index merge path.
+2). Every dnf item should exactly be used as full length index range or prefix index range, other than being not used.
+*/
+func (ds *DataSource) generateIndexMergePartialPath4Or(normalPathCnt int, indexMergeDNFConds []expression.Expression) ([]*util.AccessPath, []bool, bool, error) {
+	// step1: collect all mv index paths
+	possibleMVIndexPaths := make([]*util.AccessPath, 0, len(ds.possibleAccessPaths))
+	for idx := 0; idx < normalPathCnt; idx++ {
+		if !isMVIndexPath(ds.possibleAccessPaths[idx]) {
+			continue // not a MVIndex path
+		}
+		if !ds.isInIndexMergeHints(ds.possibleAccessPaths[idx].Index.Name.L) {
+			continue
+		}
+		possibleMVIndexPaths = append(possibleMVIndexPaths, ds.possibleAccessPaths[idx])
+	}
+	// step2: mapping index merge conditions into possible mv index path
+	mvAndPartialPaths := make([]*util.AccessPath, 0, len(possibleMVIndexPaths))
+	usedMap := make([]bool, len(indexMergeDNFConds))
+	needSelection := false
+
+	for offset, dnfCond := range indexMergeDNFConds {
+		var cnfConds []expression.Expression
+		if sf, ok := dnfCond.(*expression.ScalarFunction); !ok {
+			continue
+		} else if sf.FuncName.L == ast.LogicAnd {
+			cnfConds = expression.FlattenCNFConditions(sf)
+		} else {
+			cnfConds = []expression.Expression{sf}
+		}
```

Comment:
```suggestion
		if sf, ok := dnfCond.(*expression.ScalarFunction); !ok {
			continue
		} else {
			cnfConds = expression.FlattenCNFConditions(sf)
		}
```

Reply:
not exactly here, if a function is not ast.AND, let's say `plus(a,b)`, I want to keep it real and return as `[]expression.Expression{plus}`

while after the trails and errors, I find it will returns `[]expression.Expression{a,b}` to me, the below is the core code inside them

![image](https://github.com/pingcap/tidb/assets/15825830/10296431-afae-4293-a62c-438f39ee5949)

Reply:
You are right, my mistake. It should be `expression.SplitCNFItems()`.